core/src/main/scala/kafka/log/Log.scala

```diff
@@ -79,6 +78,17 @@ case class LogAppendInfo(var firstOffset: Long,
  *
  * @param dir The directory in which log segments are created.
  * @param config The log configuration settings
+ * @param logStartOffset The smallest offset is allowed to be exposed to kafka client.
+ *                       The logStartOffset can be updated by :
+ *                       - user's PurgeRequest
+ *                       - broker's log retention
+ *                       - broker's log truncation
```

Comment:
Discussed offline. Given that `Log.truncateTo(...)` has explicitly consider the case that the targetOffset is smaller than base offset of the first log segment, it would be safer and simpler for us to assume that `Log.truncateTo(...)` can truncate to an offset smaller than logStartOffset.